File: sdk/tools/sdkgen/internal/emit/ts/invoke_support.ts
import {
  GestaltError,
  GestaltErrorCode,
  httpStatusToGestaltCode,
} from "./rpc_support.ts";

/**
 * The canonical app invocation payload error: an HTTP-error status, an error
 * envelope, or an undecodable result body. Transport failures stay
 * GestaltError; both surface from decoded invoke methods.
 *
 * Envelope `error.code` on the wire maps to {@link InvokeError.reason}.
 */
export class InvokeError extends GestaltError {
  readonly app: string;
  readonly operation: string;
  readonly status?: number | undefined;
  readonly reason?: string | undefined;
  readonly body: unknown;
  readonly rawBody: Uint8Array;

  constructor(input: {
    app?: string | undefined;
    operation?: string | undefined;
    status?: number | undefined;
    reason?: string | undefined;
    message?: string | undefined;
    body?: unknown;
    rawBody?: Uint8Array | string | undefined;
    cause?: unknown;
  }) {
    const message = input.message ?? defaultInvokeErrorMessage(input.status);
    const gestaltCode =
      input.status != null
        ? httpStatusToGestaltCode(input.status)
        : input.reason === "graphql_errors" ||
            message === "app invoke response is not valid JSON" ||
            message === "operation result body is not valid JSON"
          ? GestaltErrorCode.Internal
          : GestaltErrorCode.Unknown;
    super(gestaltCode, message, {
      cause: input.cause,
    });
    this.name = "InvokeError";
    this.app = input.app ?? "";
    this.operation = input.operation ?? "";
    this.status = input.status;
    this.reason = input.reason;
    this.body = input.body;
    this.rawBody =
      typeof input.rawBody === "string"
        ? new TextEncoder().encode(input.rawBody)
        : new Uint8Array(input.rawBody ?? new Uint8Array());
  }

  rawText(): string {
    return new TextDecoder("utf-8", { fatal: false }).decode(this.rawBody);
  }
}

/**
 * Decodes one app operation result with the standard JSON envelope semantics:
 * success envelopes return their data, error envelopes and HTTP-error
 * statuses throw InvokeError, and any other JSON body passes through
 * unchanged.
 */
export function decodeAppResult<T = unknown>(
  app: string,
  operation: string,
  result: { status: number; body: Uint8Array },
): T {
  return decodeAppBody(app, operation, result.status, result.body) as T;
}

/**
 * Decodes one GraphQL invocation result like decodeAppResult and additionally
 * throws InvokeError when the response carries a non-empty GraphQL errors
 * array.
 */
export function decodeGraphQLResult<T = unknown>(
  app: string,
  result: { status: number; body: Uint8Array },
): T {
  const decoded = decodeAppResult<unknown>(app, "graphql", result);
  throwGraphQLErrors(app, result.body, parseOperationResultJson(result.body));
  throwGraphQLErrors(app, result.body, decoded);
  return decoded as T;
}

/**
 * Reports whether an HTTP status is a 2xx success.
 */
export function isOk(status: number): boolean {
  return status >= 200 && status <= 299;
}

/**
 * Throws InvokeError when the result's HTTP status is not a 2xx success,
 * extracting the error envelope's message and reason when the body carries
 * them. Successful results pass through untouched.
 */
export function requireOk(
  app: string,
  operation: string,
  result: { status: number; body: Uint8Array },
): void {
  if (!isOk(result.status)) {
    throw statusInvokeError(app, operation, result.status, result.body);
  }
}

function defaultInvokeErrorMessage(status?: number | undefined): string {
  return status === undefined
    ? "app invoke failed"
    : `app invoke failed with status ${status}`;
}

function throwGraphQLErrors(
  app: string,
  rawBody: Uint8Array,
  value: unknown,
): void {
  if (!isRecord(value)) {
    return;
  }
  const errors = value.errors;
  if (Array.isArray(errors) && errors.length > 0) {
    throw new InvokeError({
      app,
      operation: "graphql",
      reason: "graphql_errors",
      message: graphqlErrorMessage(errors),
      body: value,
      rawBody,
    });
  }
}

// statusInvokeError builds the canonical InvokeError for an HTTP-error
// status, extracting the error envelope's message and reason when the body
// parses as JSON.
function statusInvokeError(
  app: string,
  operation: string,
  status: number,
  body: Uint8Array,
): InvokeError {
  let parsed: unknown;
  try {
    parsed = parseOperationResultJson(body);
  } catch {
    return new InvokeError({
      app,
      operation,
      status,
      message: `app invoke failed with status ${status}`,
      rawBody: body,
    });
  }
  const fields = extractInvokeErrorFields(parsed);
  return new InvokeError({
    app,
    operation,
    status,
    reason: fields.reason,
    message: fields.message ?? `app invoke failed with status ${status}`,
    body: parsed,
    rawBody: body,
  });
}

function decodeAppBody(
  app: string,
  operation: string,
  status: number,
  body: Uint8Array,
): unknown {
  if (!isOk(status)) {
    throw statusInvokeError(app, operation, status, body);
  }
  let parsed: unknown;
  try {
    parsed = parseOperationResultJson(body);
  } catch {
    throw new InvokeError({
      app,
      operation,
      message: "app invoke response is not valid JSON",
      rawBody: body,
    });
  }

  if (isRecord(parsed) && typeof parsed.status === "string") {
    if (parsed.status === "error") {
      const fields = extractInvokeErrorFields(parsed);
      throw new InvokeError({
        app,
        operation,
        reason: fields.reason,
        message: fields.message ?? "app invoke failed",
        body: parsed,
        rawBody: body,
      });
    }
    if (parsed.status === "success" && Object.hasOwn(parsed, "data")) {
      return parsed.data;
    }
  }

  return parsed;
}

function parseOperationResultJson(body: Uint8Array): unknown {
  const text = decodeBodyText(body);
  if (text.trim() === "") {
    return {};
  }
  try {
    return JSON.parse(text) as unknown;
  } catch {
    throw new InvokeError({
      message: "operation result body is not valid JSON",
      rawBody: body,
    });
  }
}

function decodeBodyText(body: Uint8Array): string {
  return new TextDecoder("utf-8", { fatal: false }).decode(body);
}

function extractInvokeErrorFields(parsed: unknown): {
  message?: string | undefined;
  reason?: string | undefined;
} {
  if (!isRecord(parsed)) {
    return {};
  }
  const nested = isRecord(parsed.error) ? parsed.error : undefined;
  const message = stringValue(nested?.message) ?? stringValue(parsed.message);
  const reason = stringValue(nested?.code) ?? stringValue(parsed.code);
  return { message, reason };
}

function graphqlErrorMessage(errors: readonly unknown[]): string {
  const first = errors[0];
  if (
    isRecord(first) &&
    typeof first.message === "string" &&
    first.message.trim()
  ) {
    return first.message;
  }
  return "GraphQL returned errors";
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function stringValue(value: unknown): string | undefined {
  return typeof value === "string" && value.trim() ? value : undefined;
}

File: sdk/tools/sdkgen/internal/emit/python/invoke_support.py
"""Decode runtime for json_result methods: the standard JSON operation
envelope semantics shared by app invocation results. Success envelopes return
their data, error envelopes and HTTP-error statuses raise InvokeError, and any
other JSON body passes through unchanged. The requests-style ``ok`` and
``raise_for_status`` helpers expose the HTTP status semantics directly."""

from __future__ import annotations

import json
from typing import Any

from .rpc_support import GestaltError, GestaltErrorCode, JsonValue, http_status_to_gestalt_code


def _body_bytes(value: bytes | str) -> bytes:
    return value.encode("utf-8") if isinstance(value, str) else bytes(value)


class InvokeError(GestaltError):
    """Decoded app invocation failure: an HTTP-error status, an error
    envelope, or an undecodable result body.

    Envelope ``error.code`` on the wire maps to :attr:`reason`.
    """

    def __init__(
        self,
        message: str = "app invoke failed",
        *,
        app: str = "",
        operation: str = "",
        status: int | None = None,
        reason: str | None = None,
        body: JsonValue = None,
        raw_body: bytes | str = b"",
    ) -> None:
        if status is not None:
            gestalt_code = http_status_to_gestalt_code(status)
        elif reason == "graphql_errors" or message in (
            "app invoke response is not valid JSON",
            "operation result body is not valid JSON",
        ):
            gestalt_code = GestaltErrorCode.INTERNAL
        else:
            gestalt_code = GestaltErrorCode.UNKNOWN
        super().__init__(gestalt_code, message)
        self.app = app
        self.operation = operation
        self.status = status
        self.reason = reason
        self.body = body
        self.raw_body = _body_bytes(raw_body)


def ok(status: int) -> bool:
    """Report whether an HTTP status is a success (200-299), mirroring the
    requests library's ``Response.ok``."""

    return 200 <= status <= 299


def raise_for_status(app: str, operation: str, status: int, body: bytes) -> None:
    """Raise :class:`InvokeError` for a non-success HTTP status (outside
    200-299), decoding the result body for the error message and reason exactly
    like :func:`decode_app_result`. Success statuses return None, mirroring the
    requests library's ``Response.raise_for_status``."""

    if not ok(status):
        raise _http_status_error(app, operation, status, body)


def _http_status_error(
    app: str, operation: str, status: int, body: bytes
) -> InvokeError:
    """Build the :class:`InvokeError` for an HTTP-error status: a JSON body
    supplies the decoded message and reason, any other body raises on the status
    alone."""

    try:
        parsed = _parse_json_result_body(body)
    except ValueError:
        return InvokeError(
            f"app invoke failed with status {status}",
            app=app,
            operation=operation,
            status=status,
            raw_body=body,
        )
    message, reason = _message_reason_from_body(parsed)
    return InvokeError(
        message or f"app invoke failed with status {status}",
        app=app,
        operation=operation,
        status=status,
        reason=reason,
        body=parsed,
        raw_body=body,
    )


def decode_app_result(app: str, operation: str, status: int, body: bytes) -> Any:
    """Decode one app operation result with the standard JSON envelope
    semantics: success envelopes return their data, error envelopes and
    HTTP-error statuses raise :class:`InvokeError`, and any other JSON body
    passes through unchanged."""

    if not ok(status):
        raise _http_status_error(app, operation, status, body)

    try:
        parsed = _parse_json_result_body(body)
    except ValueError:
        raise InvokeError(
            "app invoke response is not valid JSON",
            app=app,
            operation=operation,
            raw_body=body,
        ) from None

    if isinstance(parsed, dict) and isinstance(parsed.get("status"), str):
        if parsed["status"] == "error":
            message, reason = _message_reason_from_body(parsed)
            raise InvokeError(
                message or "app invoke failed",
                app=app,
                operation=operation,
                reason=reason,
                body=parsed,
                raw_body=body,
            )
        if parsed["status"] == "success" and "data" in parsed:
            return parsed["data"]
    return parsed


def decode_graphql_result(app: str, status: int, body: bytes) -> Any:
    """Decode one GraphQL invocation result like :func:`decode_app_result` and
    additionally raise :class:`InvokeError` when the response carries a
    non-empty GraphQL errors array."""

    decoded = decode_app_result(app, "graphql", status, body)
    try:
        _raise_graphql_errors(app, body, _parse_json_result_body(body))
    except ValueError:
        pass
    _raise_graphql_errors(app, body, decoded)
    return decoded


def _parse_json_result_body(body: bytes) -> JsonValue:
    if body.strip() == b"":
        return {}
    return json.loads(body)


def _raise_graphql_errors(app: str, raw_body: bytes, value: JsonValue) -> None:
    if not isinstance(value, dict):
        return
    errors = value.get("errors")
    if isinstance(errors, list) and errors:
        raise InvokeError(
            _graphql_error_message(errors),
            app=app,
            operation="graphql",
            reason="graphql_errors",
            body=value,
            raw_body=raw_body,
        )


def _message_reason_from_body(value: JsonValue) -> tuple[str | None, str | None]:
    if not isinstance(value, dict):
        return None, None
    message: str | None = None
    reason: str | None = None
    error = value.get("error")
    if isinstance(error, dict):
        if isinstance(error.get("message"), str) and error["message"].strip():
            message = error["message"]
        if isinstance(error.get("code"), str) and error["code"].strip():
            reason = error["code"]
    if (
        message is None
        and isinstance(value.get("message"), str)
        and value["message"].strip()
    ):
        message = value["message"]
    if reason is None and isinstance(value.get("code"), str) and value["code"].strip():
        reason = value["code"]
    return message, reason


def _graphql_error_message(errors: list[JsonValue]) -> str:
    first = errors[0] if errors else None
    if (
        isinstance(first, dict)
        and isinstance(first.get("message"), str)
        and first["message"].strip()
    ):
        return first["message"]
    return "GraphQL returned errors"
